APS-REVAMP-01

mahendra-aps opened the pull request:
Added actor based projects core,tpp and user.

yuhanniu commented on tpp/log4net.config:
```diff
@@ -0,0 +1,11 @@
+<log4net>
```
Just wanted to double that this file is required for all core, user, tpp?

mahendra-aps replied:
Done.Removed from all core, user, tpp

yuhanniu commented on tpp/tpp.csproj:
```diff
@@ -0,0 +1,20 @@
+    <Description>C# SDK for interacting with the Token System</Description>
+    <Company>Token, Inc</Company>
+    <Authors>Token, Inc</Authors>
+    <Version>2.0.0-beta1</Version>
```
1.0.0-beta1

mahendra-aps replied:
Getting below error after changing the version to 1.0.0-beta1

System.AggregateException : One or more errors occurred. (Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK."))
---- Grpc.Core.RpcException : Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK.")

yuhanniu commented on tpp/tpp.csproj:
```diff
@@ -0,0 +1,20 @@
+ 
+ 
+  <ItemGroup>
+    <ProjectReference Include="..\core\core.csproj" />
```
Core.csproj

mahendra-aps replied:
Getting below error after changing the version to 1.0.0-beta1

System.AggregateException : One or more errors occurred. (Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK."))
---- Grpc.Core.RpcException : Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK.")

mahendra-aps replied:
done

yuhanniu commented on user/user.csproj:
```diff
@@ -0,0 +1,20 @@
+    <Description>C# SDK for interacting with the Token System</Description>
+    <Company>Token, Inc</Company>
+    <Authors>Token, Inc</Authors>
+    <Version>2.0.0-beta1</Version>
```
1.0.0-beta1

mahendra-aps replied:
Getting below error after changing the version to 1.0.0-beta1

System.AggregateException : One or more errors occurred. (Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK."))
---- Grpc.Core.RpcException : Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK.")

yuhanniu commented on user/user.csproj:
```diff
@@ -0,0 +1,20 @@
+  
+
+  <ItemGroup>
+    <ProjectReference Include="..\core\core.csproj" />
```
Core.csproj

mahendra-aps replied:
done

gianluca-token commented on user/src/TransferTokenBuilder.cs:
```diff
@@ -0,0 +1,362 @@
+			return this;
+		}
+
+		public TransferTokenBuilder From(string memberId) {
```
comment?

mahendra-aps replied:
done

mahendra-aps commented on build_proto_revamp.rb:
```diff
@@ -0,0 +1,111 @@
+else
+url = "https://token.jfrog.io/token/libs-release/#{path}/#{name}-#{type}/#{version}/#{file}"
+open(file, 'wb') do |file|
+file << open(url, :http_basic_authentication => ['vanika.appperfect', 'App4ever$']).read
```
Done

mahendra-aps commented on core/core.csproj:
```diff
@@ -0,0 +1,27 @@
+    <Description>C# SDK for interacting with the Token System</Description>
+    <Company>Token, Inc</Company>
+    <Authors>Token, Inc</Authors>
+    <Version>2.0.0-beta1</Version>
```
done

mahendra-aps commented on sdk-csharp.sln:
```diff
@@ -1,11 +1,19 @@
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Test", "tests\Test.csproj", "{23F293C9-A369-4A5A-A643-A80D932ADD7A}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "core", "core\core.csproj", "{8A173027-F8A0-4A24-B843-0349AB5107BA}"
```
done

mahendra-aps commented on sdk-csharp.sln:
```diff
@@ -1,11 +1,19 @@
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "core", "core\core.csproj", "{8A173027-F8A0-4A24-B843-0349AB5107BA}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "tpp", "tpp\tpp.csproj", "{65FADF43-4C2B-4C23-AA1C-BF0C25B207A7}"
```
done

mahendra-aps commented on sdk-csharp.sln:
```diff
@@ -1,11 +1,19 @@
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "tpp", "tpp\tpp.csproj", "{65FADF43-4C2B-4C23-AA1C-BF0C25B207A7}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "user", "user\user.csproj", "{08EF9494-B16F-444B-AF58-DFCD43DC1357}"
```
done

mahendra-aps commented on user/src/Member.cs:
```diff
@@ -0,0 +1,1462 @@
+			return GetTransfers(tokenId, offset, limit).Result;
+		}
+
+		public Task<PrepareTokenResult> PrepareTransferToken(
```
done

mahendra-aps commented on user/src/Member.cs:
```diff
@@ -0,0 +1,1462 @@
+		/// </summary>
+		/// <param name="tokenRequest">token request</param>
+		/// <returns>transfer token builder</returns>
+		public TransferTokenBuilder CreateTransferTokenBlocking(Proto.Common.TokenProtos.TokenRequest tokenRequest) {
```
Done. Changed to CreateTransferTokenBuilder

mahendra-aps commented:
> What you’ve written looks good, but I think you are missing two things:
> 
> 1. I don’t see new sample code and tests. There should be separate sample code for the tpp and user SDKs as well as separate tests for each. See user-sample and tpp-sample in the java SDK as examples.
> 2. I think initiateAccountLinking is missing? I see you copied the browser interfaces, but you’ll need to write an implementation based on some sort of webview in C# and then add initiateAccountLinking to user/Member. This is how accounts are linked in the app (since account linking requires browser logic like going to a specific URL and listening for a redirect that matches a certain regex). EDIT: Decided this does not need to be added in this iteration.

Point 1 is Completed.

yuhanniu commented:
1. so there are three `docs` folders under core/user/tpp respectively that contain basically the same files. Is there a way to not duplicate those files? There is also a `.tmp` file in each of those `docs` folder which i assume is not needed.
2. could you rebase to master and pick up the latest the changes in master and make sure they are in the new SDK too? I will try to merge the code by the end of this week, thanks.

mahendra-aps commented:
> Any update on this one? There are still merge conflicts

Done. Merge conflicts fixed.

mahendra-aps commented:
> Moreover, the `Build()` method should return the corresponding TokenClient object. For example, `Tpp.TokenClient.Builder#Build()` should return `Tpp.TokenClient`. Right now all `Build()` methods in different modules return the same `Tokenio.TokenClient` and we cast it to the right type in the tests, this is not acceptable.

I think for now C# does not support Covariant Return Types like Java. So we did this way. Please let me know if you have any other way to do this, we can use that approach.

yuhanniu commented:
I believe the [`new`](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/knowing-when-to-use-override-and-new-keywords) keyword allows you to override methods with different return types.
You are also not using generic type in the Builder class like you should.
Please we need you to try harder.
There is a test that is currently being skipped, is it being fixed?

yuhanniu commented:
@JugrajSripalJain reminded me that test was brought up by you before and I forgot to look into it. The error is because the server is unable to verify the signature that is generated for the CancelToken operation. I suggest you carefully review the code for the signature generation and then compare the final Json that's being signed with the one that's generated by the Java sdk, and see if you can find anything suspicious.

mahendra-aps commented:
> @JugrajSripalJain reminded me that test was brought up by you before and I forgot to look into it. The error is because the server is unable to verify the signature that is generated for the CancelToken operation. I suggest you carefully review the code for the signature generation, use the newest version the proto, and then compare the final Json that's being signed with the one that's generated by the Java sdk, and see if you can find anything suspicious.



> I believe the [`new`](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/knowing-when-to-use-override-and-new-keywords) keyword allows you to override methods with different return types.
> You are also not using generic type in the Builder class like you should.
> Please we need you to try harder.
> There is a test that is currently being skipped, is it being fixed?

@yuhanniu 
We are done with the changes suggested by you. Can you please have a look at those and let us know if anything else needs to be changed.
Updated the Builder class to a generic type.
Updated override methods to use different Signature type.
Also, we need your review on NewBuilder() method of Core Tokenclient class as we are getting an error here "Using the generic type 'TokenClient.Builder<T>' requires 1 type arguments". So, for now, we have removed this method in this commit as we don't see any references to this method. Is anyone going to use Core TokenClient directly?

yuhanniu commented:
I have fixed the test, refer to the PR. I have't reviewed the NewBuilder method, will probably do that tomorrow.
Since we are building a new SDK, could you also remove all the methods that are marked as obsolete? I have removed some in my commit.

mahendra-aps commented:
> I have fixed the test, refer to the [PR](https://github.com/tokenio/sdk-csharp/pull/50). I have't reviewed the NewBuilder method, will probably do that tomorrow.
> Since we are building a new SDK, could you also remove all the methods that are marked as obsolete? I have removed some in my commit.

Thanks, Yuhan. The test case is working fine on our side after this fix. We will start working on removing obsolete methods.
